File: swift/Tests/ChannelMessageTests.swift
// SPDX-License-Identifier: MIT
import XCTest
import Foundation
@testable import AetherNetProtocol

/// Unit tests for ``ChannelMessageService`` (PacketType.channelMessage). Mirrors the C#
/// `ChannelMessageTests`. A ``FakeMeshSender`` captures broadcasts — no transport needed.
///
/// The byte-identity vectors match `fixtures/channels/vectors.json` and the C# `[InlineData]`
/// vectors exactly (snake_case keys, field order channel_id, message_id, sender_uhid, content,
/// sent_at_ms, no whitespace, lowercase-dashed UUID, sent_at_ms a bare integer).
final class ChannelMessageTests: XCTestCase {

    private static let LOCAL = "aether:local:01"

    /// Mirror used only to DECODE captured ChannelMessage payloads in assertions. The real wire
    /// struct (`ChannelMessageWire`) is `private` to the service; byte-identity of the real encoder
    /// is verified separately via `_channelMessageWireBytesForTests`.
    private struct ChannelMessageWireMirror: Codable {
        let channel_id: String
        let message_id: String
        let sender_uhid: String
        let content: String
        let sent_at_ms: Int64
    }

    /// Build an inbound ChannelMessage packet, serialised via the real wire encoder.
    private func channelPacket(
        channelId: String,
        messageId: UUID,
        sender: String,
        content: String,
        sentAtMs: Int64,
        ttl: Int32 = 7
    ) -> MeshPacket {
        MeshPacket(
            type: .channelMessage,
            sourceUhid: sender,
            destinationUhid: "*",
            ttl: ttl,
            payload: _channelMessageWireBytesForTests(
                channelId: channelId,
                messageId: messageId,
                senderUhid: sender,
                content: content,
                sentAtMs: sentAtMs
            )
        )
    }

    // MARK: - Byte-identity vectors (fixtures/channels/vectors.json)

    func test_channelMessageWire_basicVector_serializesExactBytes() {
        let id = UUID(uuidString: "0f7e5d3c-1a2b-4c5d-8e9f-0a1b2c3d4e5f")!
        let data = _channelMessageWireBytesForTests(
            channelId: "res-floor-3",
            messageId: id,
            senderUhid: "aether:alice:01",
            content: "meeting at 6",
            sentAtMs: 1_700_000_000_000
        )
        let json = String(data: data, encoding: .utf8)
        XCTAssertEqual(
            json,
            "{\"channel_id\":\"res-floor-3\",\"message_id\":\"0f7e5d3c-1a2b-4c5d-8e9f-0a1b2c3d4e5f\",\"sender_uhid\":\"aether:alice:01\",\"content\":\"meeting at 6\",\"sent_at_ms\":1700000000000}"
        )
    }

    func test_channelMessageWire_minimalVector_serializesExactBytes() {
        let id = UUID(uuidString: "00000000-0000-0000-0000-000000000000")!
        let data = _channelMessageWireBytesForTests(
            channelId: "g",
            messageId: id,
            senderUhid: "n",
            content: "",
            sentAtMs: 0
        )
        let json = String(data: data, encoding: .utf8)
        XCTAssertEqual(
            json,
            "{\"channel_id\":\"g\",\"message_id\":\"00000000-0000-0000-0000-000000000000\",\"sender_uhid\":\"n\",\"content\":\"\",\"sent_at_ms\":0}"
        )
    }

    // MARK: - Publish

    func test_publish_broadcastsChannelMessage() async {
        let sender = FakeMeshSender(localUhid: "aether:alice:01")
        let svc = ChannelMessageService(sender: sender)

        _ = await svc.publish("res-floor-3", content: "meeting at 6")

        let broadcasts = sender.broadcasts()
        XCTAssertEqual(broadcasts.count, 1)
        let pkt = broadcasts[0]
        XCTAssertEqual(pkt.type, .channelMessage)
        XCTAssertEqual(pkt.destinationUhid, "*")
        XCTAssertEqual(pkt.ttl, ProtocolConstants.defaultTtl)

        let body = try? JSONDecoder().decode(ChannelMessageWireMirror.self, from: pkt.payload)
        XCTAssertEqual(body?.channel_id, "res-floor-3")
        XCTAssertEqual(body?.content, "meeting at 6")
        XCTAssertEqual(body?.sender_uhid, "aether:alice:01")
    }

    // MARK: - Handle

    func test_handle_subscribedChannel_raisesEvent() async {
        let svc = ChannelMessageService(sender: FakeMeshSender(localUhid: Self.LOCAL))
        await svc.subscribe("res-floor-3")

        let got = Locked<ChannelMessageReceived?>(nil)
        await svc.setOnMessageReceived { e in got.value = e }

        let ok = await svc.handle(channelPacket(
            channelId: "res-floor-3",
            messageId: UUID(),
            sender: "aether:bob:02",
            content: "hello floor",
            sentAtMs: 1_700_000_000_000
        ))

        XCTAssertTrue(ok)
        XCTAssertNotNil(got.value)
        XCTAssertEqual(got.value?.channelId, "res-floor-3")
        XCTAssertEqual(got.value?.content, "hello floor")
        XCTAssertEqual(got.value?.senderUhid, "aether:bob:02")
    }

    func test_handle_unsubscribedChannel_noEventButProcessed() async {
        let svc = ChannelMessageService(sender: FakeMeshSender(localUhid: Self.LOCAL))
        let raised = Locked<Bool>(false)
        await svc.setOnMessageReceived { _ in raised.value = true }

        let ok = await svc.handle(channelPacket(
            channelId: "society-x",
            messageId: UUID(),
            sender: "aether:bob:02",
            content: "hi",
            sentAtMs: 1
        ))

        XCTAssertTrue(ok)              // processed + relayed
        XCTAssertFalse(raised.value)   // but not surfaced — we aren't subscribed
    }

    func test_handle_duplicateMessageId_returnsFalse() async {
        let svc = ChannelMessageService(sender: FakeMeshSender(localUhid: Self.LOCAL))
        await svc.subscribe("res-floor-3")
        let id = UUID()

        let events = Locked<Int>(0)
        await svc.setOnMessageReceived { _ in events.value = events.value + 1 }

        let first = await svc.handle(channelPacket(
            channelId: "res-floor-3", messageId: id, sender: "aether:bob:02", content: "one", sentAtMs: 1))
        let second = await svc.handle(channelPacket(
            channelId: "res-floor-3", messageId: id, sender: "aether:bob:02", content: "one", sentAtMs: 1))

        XCTAssertTrue(first)
        XCTAssertFalse(second)
        XCTAssertEqual(events.value, 1)
    }

    func test_handle_wrongPacketType_returnsFalse() async {
        let svc = ChannelMessageService(sender: FakeMeshSender(localUhid: Self.LOCAL))
        var pkt = channelPacket(
            channelId: "res-floor-3", messageId: UUID(), sender: "aether:bob:02", content: "x", sentAtMs: 1)
        pkt.type = .data
        let ok = await svc.handle(pkt)
        XCTAssertFalse(ok)
    }

    func test_handle_relaysWhenTtlAllows() async {
        let relaySender = FakeMeshSender(localUhid: "aether:relay:09")
        let svc = ChannelMessageService(sender: relaySender) // not subscribed — pure relay
        _ = await svc.handle(channelPacket(
            channelId: "res-floor-3", messageId: UUID(), sender: "aether:bob:02", content: "hop", sentAtMs: 1, ttl: 5))

        let broadcasts = relaySender.broadcasts()
        XCTAssertEqual(broadcasts.count, 1)
        XCTAssertEqual(broadcasts[0].type, .channelMessage)
        XCTAssertEqual(broadcasts[0].ttl, 4)
    }

    func test_handle_ownMessage_notSurfacedAndNotRelayed() async {
        // A node's own message flooding back: subscribed to the channel, but sender == local.
        let sender = FakeMeshSender(localUhid: "aether:me:01")
        let svc = ChannelMessageService(sender: sender)
        await svc.subscribe("res-floor-3")
        let raised = Locked<Bool>(false)
        await svc.setOnMessageReceived { _ in raised.value = true }

        let ok = await svc.handle(channelPacket(
            channelId: "res-floor-3", messageId: UUID(), sender: "aether:me:01", content: "mine", sentAtMs: 1, ttl: 5))

        XCTAssertTrue(ok)                       // de-dup slot consumed → processed
        XCTAssertFalse(raised.value)            // not surfaced — it is our own
        XCTAssertEqual(sender.broadcasts().count, 0) // not relayed — it is our own
    }

    // MARK: - Subscriptions

    func test_subscriptions_trackAndUntrack() async {
        let svc = ChannelMessageService(sender: FakeMeshSender(localUhid: Self.LOCAL))
        await svc.subscribe("a")
        await svc.subscribe("b")
        var subs = await svc.getSubscriptions()
        XCTAssertEqual(Set(subs), ["a", "b"])

        await svc.unsubscribe("a")
        subs = await svc.getSubscriptions()
        XCTAssertEqual(Set(subs), ["b"])
    }
}

File: tests/AetherNet.Core.Tests/ChannelMessageTests.cs
// SPDX-License-Identifier: MIT

using System.Text.Json;
using AetherNet.Channels;
using AetherNet.Models;
using AetherNet.Protocol;
using AetherNet.Routing;
using Microsoft.Extensions.Logging.Abstractions;
using Xunit;

namespace AetherNet.Core.Tests;

/// <summary>
/// Unit tests for <see cref="ChannelMessageService"/> (PacketType.ChannelMessage). Fake
/// <see cref="IMeshSender"/> captures broadcasts — no transport needed.
/// </summary>
public sealed class ChannelMessageTests
{
    private sealed class FakeMeshSender : IMeshSender
    {
        public string LocalUhid { get; set; } = "aether:local:01";
        public List<MeshPacket> Broadcasts { get; } = [];

        public IReadOnlyList<PeerInfo> GetConnectedPeers() => Array.Empty<PeerInfo>();
        public Task<bool> SendAsync(MeshPacket packet, string nextHopUhid, CancellationToken ct = default)
            => Task.FromResult(true);
        public Task<int> BroadcastAsync(MeshPacket packet, CancellationToken ct = default)
        {
            Broadcasts.Add(packet);
            return Task.FromResult(1);
        }
    }

    private static readonly JsonSerializerOptions JsonOpts = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
    };

    private static ChannelMessageService Build(FakeMeshSender sender)
        => new(sender, NullLogger<ChannelMessageService>.Instance);

    private static MeshPacket ChannelPacket(
        string channelId, Guid messageId, string sender, string content, long sentAtMs, int ttl = 7) => new()
    {
        Type = PacketType.ChannelMessage,
        SourceUhid = sender,
        DestinationUhid = "*",
        Ttl = ttl,
        Payload = JsonSerializer.SerializeToUtf8Bytes(new ChannelMessagePayload
        {
            ChannelId = channelId,
            MessageId = messageId,
            SenderUhid = sender,
            Content = content,
            SentAtMs = sentAtMs,
        }, JsonOpts),
    };

    [Theory]
    [InlineData("res-floor-3", "0f7e5d3c-1a2b-4c5d-8e9f-0a1b2c3d4e5f", "aether:alice:01", "meeting at 6", 1700000000000L,
        "{\"channel_id\":\"res-floor-3\",\"message_id\":\"0f7e5d3c-1a2b-4c5d-8e9f-0a1b2c3d4e5f\",\"sender_uhid\":\"aether:alice:01\",\"content\":\"meeting at 6\",\"sent_at_ms\":1700000000000}")]
    [InlineData("g", "00000000-0000-0000-0000-000000000000", "n", "", 0L,
        "{\"channel_id\":\"g\",\"message_id\":\"00000000-0000-0000-0000-000000000000\",\"sender_uhid\":\"n\",\"content\":\"\",\"sent_at_ms\":0}")]
    public void ChannelMessagePayload_SerializesToCanonicalBytes(
        string channelId, string messageId, string sender, string content, long sentAtMs, string expected)
    {
        var payload = new ChannelMessagePayload
        {
            ChannelId = channelId,
            MessageId = Guid.Parse(messageId),
            SenderUhid = sender,
            Content = content,
            SentAtMs = sentAtMs,
        };
        var json = System.Text.Encoding.UTF8.GetString(JsonSerializer.SerializeToUtf8Bytes(payload, JsonOpts));
        Assert.Equal(expected, json);
    }

    [Fact]
    public async Task Publish_BroadcastsChannelMessage()
    {
        var sender = new FakeMeshSender { LocalUhid = "aether:alice:01" };
        var svc = Build(sender);

        await svc.PublishAsync("res-floor-3", "meeting at 6");

        var pkt = Assert.Single(sender.Broadcasts);
        Assert.Equal(PacketType.ChannelMessage, pkt.Type);
        var body = JsonSerializer.Deserialize<ChannelMessagePayload>(pkt.Payload, JsonOpts)!;
        Assert.Equal("res-floor-3", body.ChannelId);
        Assert.Equal("meeting at 6", body.Content);
        Assert.Equal("aether:alice:01", body.SenderUhid);
    }

    [Fact]
    public async Task Handle_SubscribedChannel_RaisesEvent()
    {
        var svc = Build(new FakeMeshSender { LocalUhid = "aether:local:01" });
        svc.Subscribe("res-floor-3");

        ChannelMessageReceived? got = null;
        svc.MessageReceived += (_, e) => got = e;

        var ok = await svc.HandleAsync(
            ChannelPacket("res-floor-3", Guid.NewGuid(), "aether:bob:02", "hello floor", 1700000000000L));

        Assert.True(ok);
        Assert.NotNull(got);
        Assert.Equal("res-floor-3", got!.ChannelId);
        Assert.Equal("hello floor", got.Content);
        Assert.Equal("aether:bob:02", got.SenderUhid);
    }

    [Fact]
    public async Task Handle_UnsubscribedChannel_NoEventButProcessed()
    {
        var svc = Build(new FakeMeshSender { LocalUhid = "aether:local:01" });
        var raised = false;
        svc.MessageReceived += (_, _) => raised = true;

        var ok = await svc.HandleAsync(
            ChannelPacket("society-x", Guid.NewGuid(), "aether:bob:02", "hi", 1L));

        Assert.True(ok);     // processed + relayed
        Assert.False(raised); // but not surfaced — we aren't subscribed
    }

    [Fact]
    public async Task Handle_DuplicateMessageId_ReturnsFalse()
    {
        var svc = Build(new FakeMeshSender { LocalUhid = "aether:local:01" });
        svc.Subscribe("res-floor-3");
        var id = Guid.NewGuid();

        var events = 0;
        svc.MessageReceived += (_, _) => events++;

        Assert.True(await svc.HandleAsync(ChannelPacket("res-floor-3", id, "aether:bob:02", "one", 1L)));
        Assert.False(await svc.HandleAsync(ChannelPacket("res-floor-3", id, "aether:bob:02", "one", 1L)));
        Assert.Equal(1, events);
    }

    [Fact]
    public async Task Handle_WrongPacketType_ReturnsFalse()
    {
        var svc = Build(new FakeMeshSender());
        var pkt = ChannelPacket("res-floor-3", Guid.NewGuid(), "aether:bob:02", "x", 1L);
        pkt.Type = PacketType.Data;
        Assert.False(await svc.HandleAsync(pkt));
    }

    [Fact]
    public async Task Handle_RelaysWhenTtlAllows()
    {
        var relaySender = new FakeMeshSender { LocalUhid = "aether:relay:09" };
        var svc = Build(relaySender); // not subscribed — pure relay
        await svc.HandleAsync(ChannelPacket("res-floor-3", Guid.NewGuid(), "aether:bob:02", "hop", 1L, ttl: 5));

        var relayed = Assert.Single(relaySender.Broadcasts);
        Assert.Equal(PacketType.ChannelMessage, relayed.Type);
        Assert.Equal(4, relayed.Ttl);
    }
}
